<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
    <meta content="text/html; charset=windows-1252"
      http-equiv="content-type">
    <title>adegenet on the web</title>
    <script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-20083187-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
  </script>
  </head>
  <body>
    <div style="margin-left: 40px; text-align: center;"> <img
        style="width: 600px; height: 90px;" alt="title"
        src="images/title.png"><br>
      <br>
      <br>
      <br>
      <div style="text-align: left;"> <span style="font-weight: bold;">adegenet</span>
        is an <span style="text-decoration: underline;"><a
            href="http://www.r-project.org/" target="_top"><img alt=""
              src="images/R.png" style="border: 0px solid ; width: 30px;
              height: 23px;"></a></span> package dedicated to the
        exploratory analysis of genetic data. It implements a set of
        tools ranging from multivariate methods to spatial genetics and
        genome-wise SNP data analysis. <br>
        <br>
        It is developed on <a href="http://sourceforge.net/p/adegenet">Sourceforge</a>
        by <a href="http://sites.google.com/site/thibautjombart/">Thibaut





          Jombart</a>, <a
          href="https://sites.google.com/site/caitlincollinswebpage/home">Caitlin
          Collins</a>, Ismail Ahmed, <a
          href="http://www.federicocalboli.com/">Federico Calboli</a>, <a
          href="http://www1.imperial.ac.uk/medicine/people/a.cori/">Anne
          Cori</a>, Tobias Erik Reiners, and <a
          href="https://sites.google.com/site/psolymos/">Péter Sólymos</a>,
        and officially released on <a href="http://cran.r-project.org">CRAN</a>
        periodically. <br>
        <br>
        adegenet is described in the following publications:<br>
        <ul>
          <li> Jombart T. (2008) adegenet: a R package for the
            multivariate analysis of genetic markers. <span
              style="font-style: italic;">Bioinformatics</span> <span
              style="font-weight: bold;">24</span>: 1403-1405. doi:
            10.1093/bioinformatics/btn129 [<a
href="http://bioinformatics.oxfordjournals.org/cgi/reprint/btn129?ijkey=6sqx5BTXCdYtBZz&amp;keytype=ref">link</a>
            to a free pdf]<br>
            <br>
          </li>
          <li>Jombart T. and Ahmed I. (2011) <span style="font-style:
              italic;">adegenet 1.3-1</span>: new tools for the analysis
            of genome-wide SNP data. <span style="font-style: italic;">Bioinformatics</span>.
            doi: 10.1093/bioinformatics/btr521 [<a
href="http://bioinformatics.oxfordjournals.org/content/early/2011/09/16/bioinformatics.btr521">link
to








              the bublisher's website</a>]<br>
          </li>
        </ul>
        <br>
        <div style="text-align: center;"> <img style="width: 600px;
            height: 451px;" alt="" src="images/acceuil.png" height="788"
            width="1044"><br>
          <br>
        </div>
        <div style="text-align: left;">
          <div style="text-align: center;"> <small>sPCA, DAPC,
              typological coherence of markers, Monmonier&nbsp;
              algorithm, ...<br>
              <br>
            </small></div>
          <small><br>
          </small><span style="text-decoration: underline;">Main
            features of adegenet are:</span><br>
          <ul>
            <li>data representation (<span style="font-weight: bold;">classes</span>)
              suitable for multivariate analysis</li>
            <li>data <span style="font-weight: bold;">import</span>
              from GENETIX, STRUCTURE, Genepop, Fstat, Easypop, or any
              dataframe of genotypes</li>
            <li>data import from <span style="font-weight: bold;">aligned


                DNA sequences</span> to <span style="font-weight:
                bold;">SNPs</span></li>
            <li> data import from <span style="font-weight: bold;">aligned



                protein sequences</span> to polymorphic sites </li>
            <li> data <span style="font-weight: bold;">export</span> to
              the R packages genetics, hierfstat, LDheatmap</li>
            <li> handling of <span style="font-weight: bold;">different
                levels of ploidy</span></li>
            <li>handling of <span style="font-weight: bold;">codominant
              </span>markersand<span style="font-weight: bold;">presence/absence</span>
              data</li>
            <li>basic and advanced <span style="font-weight: bold;">data








                manipulation</span></li>
            <li> basic <span style="font-weight: bold;">data
                information </span>(heterozygosity, numbers of alleles,
              sample sizes, ...)</li>
            <li> <span style="font-weight: bold;">HWE</span> and <span
                style="font-weight: bold;">G-statistic</span> <span
                style="font-weight: bold;">tests</span>, F statistics
              implemented for adegenet objects</li>
            <li> computation of <span style="font-weight: bold;">genetic</span>
              <span style="font-weight: bold;">distances</span></li>
            <li>computation of <span style="font-weight: bold;">pairwise








                Fst </span></li>
            <li> simulation of <span style="font-weight: bold;">hybridization</span></li>
            <li>methods for <span style="font-weight: bold;">spatial
                genetics: sPCA, </span><span style="font-weight: bold;">tests


                for global and local structuring, </span><span
                style="font-weight: bold;">Monmonier algorithm</span></li>
            <li>the <span style="font-style: italic; font-weight:
                bold;">seqTrack</span><span style="font-weight: bold;">algorithm








              </span>for reconstructing genealogies of haplotypes</li>
            <li>simulation of <span style="font-weight: bold;">genealogies








                of haplotypes</span></li>
            <li> Discriminant Analysis of Principal Components (<span
                style="font-weight: bold;">DAPC</span>)</li>
            <li> efficient genome-wise SNP data handling and analysis </li>
            <li> <b>extraction of SNPs</b> from genomic alignments </li>
            <li> <span style="font-weight: bold;">graph-based
                clustering</span> of genomic data</li>
            <li> identification of mutations between pairs of sequences
            </li>
            <li><b>visualization of SNPs density</b> and test for the
              randomness of their distribution<img style="width: 80px;
                height: 37px;" alt="" src="images/new.png" height="37"
                width="80"></li>
            <li><b>web interface for DAPC</b>, including feature
              selection and cross validation <img style="width: 80px;
                height: 37px;" alt="" src="images/new.png" height="37"
                width="80"><br>
            </li>
          </ul>
        </div>
        <span style="font-weight: bold;"><br>
          <span style="text-decoration: underline; font-weight: bold;">Maintainer</span></span><span
          style="text-decoration: underline; font-weight: bold;">:</span>
        Thibaut Jombart (<a target="_new"
          href="http://sites.google.com/site/thibautjombart/">website</a>)<br>
        <span style="text-decoration: underline;">Developers:</span>
        Thibaut Jombart (<a href="mailto:tjombart@imperial.ac.uk">tjombart@imperial.ac.uk</a>),
Caitlin



        Collins (<a href="mailto:caitiecollins17@gmail.com">caitiecollins@gmail.com</a>),








        Ismaïl Ahmed (<a href="mailto:ismail.ahmed@inserm.fr">ismail.ahmed@inserm.fr</a>),



        Federico Calboli (<a href="mailto:f.calboli@imperial.ac.uk">f.calboli@imperial.ac.uk</a>),






        Anne Cori (<a href="mailto:a.cori@imperial.ac.uk">a.cori@imperial.ac.uk</a>),
Tobias








        Erik Reiners (<a href="mailto:Tobias.Reiners@Senckenberg.de">Tobias.Reiners@Senckenberg.de</a>),








        Péter Sólymos (<a href="mailto:solymos@ualberta.ca">solymos@ualberta.ca</a>)<br>
        <span style="text-decoration: underline;">Contributors
          (datasets/ideas):</span> Christophe Fraser, Katayoun
        Moazami-Goudarzi, Denis Laloë, Francois Balloux, Dominique
        Pontier, Daniel Maillard<span style="font-weight: bold;"><br>
          <br>
          <br style="font-weight: bold;">
          Suggestions, comments and contributions are most welcome!</span><br>
        <br>
        <br>
        <br>
        <span style="font-weight: bold;"></span></div>
    </div>
  </body>
</html>
