<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
    <meta content="text/html; charset=windows-1252"
      http-equiv="content-type">
    <title>documentation</title>
    <script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-20083187-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
  </script>
  </head>
  <body>
    <br>
    <br>
    <br>
    <div style="margin-left: 40px;"><span style="font-family:
        Helvetica,Arial,sans-serif;">As a </span><span
        style="text-decoration: underline; font-family:
        Helvetica,Arial,sans-serif;"><a href="http://www.r-project.org/"
          target="_top"><img alt="" src="images/R.png" style="border:
            0px solid ; width: 30px; height: 23px;"></a></span><span
        style="font-family: Helvetica,Arial,sans-serif;"> package, each
        component of adegenet is documented in help pages. As a starting
        point, one can type </span><span style="font-weight: bold;
        font-family: Helvetica,Arial,sans-serif;">?adegenet</span><span
        style="font-family: Helvetica,Arial,sans-serif;"> in R to get an
        overview of package. Other sources of information distributed
        with the package include a manual and vignettes. This website
        also provides other resources such as lectures and practicals.</span><br
        style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <span style="font-weight: bold; font-family:
        Helvetica,Arial,sans-serif;">To search quickly</span><span
        style="font-family: Helvetica,Arial,sans-serif;"> through these
        resources, use the </span><a style="font-family:
        Helvetica,Arial,sans-serif;" target="_top" href="search.html">search


































        engine</a><span style="font-family: Helvetica,Arial,sans-serif;">.
      </span><br style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <span style="font-weight: bold; color: rgb(0, 102, 0);
        font-family: Helvetica,Arial,sans-serif;">Manuals</span><br
        style="font-family: Helvetica,Arial,sans-serif;">
      <span style="font-family: Helvetica,Arial,sans-serif;"> - the
        adegenet </span><span style="font-weight: bold; font-family:
        Helvetica,Arial,sans-serif;">manual</span><span
        style="font-family: Helvetica,Arial,sans-serif;"> (</span><a
        style="font-family: Helvetica,Arial,sans-serif;"
        href="http://cran.r-project.org/web/packages/adegenet/adegenet.pdf">pdf</a><span
        style="font-family: Helvetica,Arial,sans-serif;">), which
        compiles all the manpages of the package.</span><br
        style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <span style="color: rgb(0, 102, 0); font-weight: bold;
        font-family: Helvetica,Arial,sans-serif;">Tutorials*</span>
      <ul>
        <li><a style="font-family: Helvetica,Arial,sans-serif;"
            href="files/tutorial-basics.pdf"><span style="font-style:
              italic;">tutorial-basics</span></a><span
            style="font-family: Helvetica,Arial,sans-serif;">:
            introduction to the package: data classes, inputs/outputs,
            basic population genetics, and exploratory analyses;
            accessible from R by typing </span><span style="color:
            rgb(255, 0, 0); font-family: Helvetica,Arial,sans-serif;"></span><span
            style="font-family: monospace; color: rgb(255, 0, 0);">adegenetTutorial("basics")</span>.</li>
        <li><a style="font-family: Helvetica,Arial,sans-serif;"
            href="files/tutorial-dapc.pdf"><span style="font-style:
              italic;">tutorial-dapc</span></a><span style="font-family:
            Helvetica,Arial,sans-serif;">: a tutorial for the analysis
            of population diversity using Discriminant Analysis of
            Principal Components (DAPC); accessible from R by typing</span>&nbsp;<span
            style="font-family: monospace; color: rgb(255, 0, 0);"><span
              style="font-family: monospace; color: rgb(255, 0, 0);">adegenetTutorial</span>("dapc")</span>.</li>
        <li><a style="font-family: Helvetica,Arial,sans-serif;"
            href="files/tutorial-genomics.pdf"><span style="font-style:
              italic;">tutorial-genomics</span></a><span
            style="font-family: Helvetica,Arial,sans-serif;">:
            introduction to the genlight class for storing and handling
            large genome-wide SNP data; accessible from R by typing</span>
          <span style="font-family: monospace;"></span><span
            style="font-family: monospace; color: rgb(255, 0, 0);">adegenetTutorial("genomics")</span>.</li>
        <li><a style="font-family: Helvetica,Arial,sans-serif;"
            href="files/tutorial-spca.pdf"><span style="font-style:
              italic;">tutorial-spca</span></a><span style="font-family:
            Helvetica,Arial,sans-serif;">: introduction to spatial
            multivariate analysis using the spatial Principal Component
            Analysis (sPCA); accessible from R by typing</span>&nbsp;<span
            style="font-family: monospace; color: rgb(255, 0, 0);">adegenetTutorial</span><span
            style="font-family: monospace; color: rgb(255, 0, 0);">("spca")</span>.</li>
      </ul>
    </div>
    <div style="margin-left: 40px;"> <span style="color: rgb(0, 102,
        0);"><small>*(note: since adegenet 1.4-0, these tutorials
          replace the vignettes which used to be part of the package)</small></span><br>
      &nbsp;&nbsp;&nbsp;&nbsp; <br>
      <span style="font-weight: bold; color: rgb(0, 102, 0);
        font-family: Helvetica,Arial,sans-serif;"><br>
        Lectures and practicals</span><br style="font-family:
        Helvetica,Arial,sans-serif;">
      <ul style="font-family: Helvetica,Arial,sans-serif;">
        <li>&nbsp; globaldiv summer school 2009, Piacenza, Italie: <a
            href="files/globaldiv-TJ.1.3.pdf">talk</a> - <a
            href="files/globaldiv-PC.1.3.pdf">practical course<br>
            <br>
          </a></li>
        <li>&nbsp;Master 2 "Ecology, Evolution, Biometry", Lyon, France:
          <a href="files/talk-E2B.1.0.pdf">talk</a> - <a
            href="files/TD-E2B.pdf">practical course<br>
            <br>
          </a></li>
        <li>One-week course on Genetic Data Analysis, Bogota, Colombia:</li>
        <ul>
          <li><a href="files/day1.1.2.pdf">lecture 1</a>: methodological
            overview</li>
          <li><a href="files/day2.1.3.pdf">lecture 2</a>: finding
            clusters, using cluster information</li>
          <li><a href="files/day3.1.2.pdf">lecture 3</a>: spatial
            genetics</li>
          <li><a href="files/practical-day1.1.2.pdf">practical 1</a> [<a
              href="files/practical-day1.1.2-allcmd.pdf">version with
              all commands</a>]</li>
          <li><a href="files/practical-day2.1.3.pdf">practical 2</a> [<a
              href="files/practical-day2.1.3-allcmd.pdf">version with
              all commands</a>]</li>
          <li><a href="files/practical-day3.1.2.pdf">practical 3</a>&nbsp;</li>
        </ul>
      </ul>
      <ul style="font-family: Helvetica,Arial,sans-serif;">
        <li>MRC methodological session: Introduction to phylogenetics:</li>
        <ul>
          <li><a href="files/MRCsession2-talk.1.2.pdf">slides</a></li>
          <li><a href="files/MRC-session2-tuto.1.3.pdf">practical</a></li>
          <li>data (<a href="files/usflu.fasta">DNA sequences</a>, <a
              href="files/usflu.annot.csv">annotations</a>)<br>
          </li>
        </ul>
      </ul>
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <ul style="font-family: Helvetica,Arial,sans-serif;">
        <li> RAPID-NGS workshop, Münster (March 2013) - <span
            style="font-weight: bold;">Outbreak analysis using <span
              style="font-style: italic;">ape </span>and</span> <span
            style="font-weight: bold;"><span style="font-style: italic;">adegenet</span></span></li>
        <ul>
          <li><a href="files/seqTrack-practical.1.3.pdf">practical</a></li>
          <li>data (<a href="files/fakeOutbreak/cases.csv">cases.csv</a>,
            <a href="files/fakeOutbreak/alignment.fa">alignment.fa</a>,
            <a href="files/fakeOutbreak/update.csv">update.csv</a>)<br>
          </li>
        </ul>
      </ul>
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <ul style="font-family: Helvetica,Arial,sans-serif;">
        <li>Imperial College - Master <span style="font-style: italic;">Modern

































            Epidemiology</span> / <i>Public Health</i>: <span
            style="font-weight: bold;">Introduction to phylogenetics</span>
          (last updated on the 15/13/2013)</li>
        <ul>
          <li><a href="files/MSc-intro-phylo.1.1.pdf">practical</a></li>
          <li>data (<a href="files/usflu.fasta">usflu.fasta</a>, <a
              href="files/usflu.annot.csv">usflu.annot.csv</a>)<br>
          </li>
        </ul>
      </ul>
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <ul style="font-family: Helvetica,Arial,sans-serif;">
        <li>Imperial College - Master <span style="font-style: italic;">Modern

































            Epidemiology</span> / <i>Public Health</i>: <span
            style="font-weight: bold;">Introduction to multivariate
            analysis / bacterial GWAS</span> (last updated on the
          21/13/2013)</li>
        <ul>
          <li><a href="files/simGWAS/lecture-MSc-MVA.1.2.pdf">lecture</a></li>
          <li><a href="files/simGWAS/MSc-GWAS-1.1.pdf">practical</a></li>
          <li>data (<a href="files/simGWAS/simGWAS.RData">simGWAS.RData</a>)<br>
          </li>
        </ul>
      </ul>
      <ul style="font-family: Helvetica,Arial,sans-serif;">
        <li>&nbsp; GenEpi workshop, London (April 2013) - <span
            style="font-weight: bold;">Outbreak analysis using <span
              style="font-style: italic;">ape, adegenet,</span> </span><b>and


























          </b><b><i>outbreaker</i></b></li>
        <ul>
          <li><a href="files/GenEpiTalk.1.0.pdf">talk</a></li>
          <li><a href="files/outbreakPractical.1.3.pdf">practical</a></li>
          <li>data (<a href="files/fakeOutbreak/cases.csv">cases.csv</a>,
            <a href="files/fakeOutbreak/alignment.fa">alignment.fa</a>,
            <a href="files/fakeOutbreak/update.csv">update.csv</a>)<br>
          </li>
        </ul>
      </ul>
      <ul style="font-family: Helvetica,Arial,sans-serif;">
        <li>&nbsp; Molecular Outbreak Epidemiology workshop, Stockholm,
          Sweden (November 2013) - <span style="font-weight: bold;">Outbreak















            analysis using <span style="font-style: italic;">ape,
              adegenet,</span> </span><b>and </b><b><i>outbreaker</i></b></li>
        <ul>
          <li>talk</li>
          <li><a href="files/outbreakPractical.1.4.pdf">practical</a></li>
          <li>data (<a href="files/fakeOutbreak/cases.csv">cases.csv</a>,
            <a href="files/fakeOutbreak/alignment.fa">alignment.fa</a>,
            <a href="files/fakeOutbreak/update.csv">update.csv</a>)<br>
          </li>
        </ul>
      </ul>
      <ul style="font-family: Helvetica,Arial,sans-serif;">
        <li>Workshop "<span style="font-weight: bold;">Phylogenetics and
            population genetics using R</span>", Montpellier (March
          2014)&nbsp;</li>
        <ul>
          <li>Multivariate analysis of genetic data: an introduction [<a
              href="files/montpellier/lecture-MVA.1.0.pdf">lecture</a>,
            <a href="files/montpellier/practical-MVAintro.1.0.pdf">practical</a>]
          </li>
          <li>Multivariate analysis of genetic data: exploring group
            diversity [<a
              href="files/montpellier/lecture-MVAgroups.1.0.pdf">lecture</a>,
            <a href="files/montpellier/practical-MVAgroups.1.0.pdf">practical</a>]<br>
          </li>
          <li>Multivariate analysis of genetic data: uncovering spatial
            structures [<a
              href="files/montpellier/lecture-MVAspatial.1.0.pdf">lecture</a>,
            <a href="files/montpellier/practical-MVAspatial.1.0.pdf">practical</a>]<br>
            <br>
          </li>
        </ul>
        <li>XXIV International Symposium of Statistics, Bogota, Colombia
          (July 2014) - <span style="font-weight: bold;">Multivariate
            analysis of genetic data: an introduction for statisticians</span>&nbsp;</li>
        <ul>
          <li><a href="files/lectureBogota2014-MVA.1.1.pdf">slides</a>
            of the lecture<br>
            <br>
          </li>
        </ul>
        <li>Workshop "<span style="font-weight: bold;">Genetic data
            analysis using R</span>" - Leuven (October 2014)&nbsp;<img
            alt="" src="images/new.png" height="37" width="80"></li>
        <ul>
          <li>Short R refresher [<a
              href="files/Leuven2014/Refresh-R.pdf">lecture</a>]<br>
          </li>
          <li>Introduction to phylogenetics [<a
              href="files/Leuven2014/lecture-phylo.1.3.pdf">lecture</a>,
            <a href="files/Leuven2014/practical-introphylo.1.1.pdf">practical</a>]
          </li>
          <li>Multivariate analysis of genetic data: an introduction [<a
              href="files/Leuven2014/lecture-MVA.1.1.pdf">lecture</a>, <a
              href="files/Leuven2014/practical-MVAintro.1.0.pdf">practical</a>]
          </li>
          <li>Multivariate analysis of genetic data: exploring group
            diversity [<a
              href="files/Leuven2014/lecture-MVAgroups.1.1.pdf">lecture</a>,
            <a href="files/Leuven2014/practical-MVAgroups.1.2.pdf">practical</a>]</li>
          <li>An introduction to Genome Wise Association Studies (GWAS)
            [lecture, practical]</li>
          <li>Multivariate analysis of genetic data: uncovering spatial
            structures [lecture, practical]<br>
            <br>
          </li>
        </ul>
        <ul>
        </ul>
      </ul>
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <span style="font-weight: bold; color: rgb(0, 102, 0);
        font-family: Helvetica,Arial,sans-serif;">Dirty hacks useful for
        teaching</span><br style="font-family:
        Helvetica,Arial,sans-serif;">
      <span style="font-family: Helvetica,Arial,sans-serif;"> - a small
        hack for installing R packages on computers without
        administrative rights: </span><a style="font-family:
        Helvetica,Arial,sans-serif;" href="files/hackLib/hackLib.R">hackLib.R</a><span
        style="font-family: Helvetica,Arial,sans-serif;">&nbsp; </span><a
        style="font-family: Helvetica,Arial,sans-serif;"
        href="files/hackLib/README">README</a> <br style="font-family:
        Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <span style="font-weight: bold; color: rgb(0, 102, 0);
        font-family: Helvetica,Arial,sans-serif;">Mailing lists archives
      </span><br style="font-family: Helvetica,Arial,sans-serif;">
      <span style="font-family: Helvetica,Arial,sans-serif;"> Use the </span><a
        style="font-family: Helvetica,Arial,sans-serif;" target="_top"
        href="search.html">search engine</a><span style="font-family:
        Helvetica,Arial,sans-serif;"> to browse the </span><span
        style="font-weight: bold; font-family:
        Helvetica,Arial,sans-serif;">archives</span><span
        style="font-family: Helvetica,Arial,sans-serif;"> of the </span><span
        style="font-weight: bold; font-family:
        Helvetica,Arial,sans-serif;">adegenet forum</span><span
        style="font-family: Helvetica,Arial,sans-serif;">.</span><br
        style="font-family: Helvetica,Arial,sans-serif;">
      <span style="font-family: Helvetica,Arial,sans-serif;"> See the '</span><a
        style="font-family: Helvetica,Arial,sans-serif;"
        href="contact.html">contact</a><span style="font-family:
        Helvetica,Arial,sans-serif;">' section of this website for
        further information on mailing lists.</span><br
        style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <span style="font-weight: bold; color: rgb(0, 102, 0);
        font-family: Helvetica,Arial,sans-serif;">Miscellaneous<br>
      </span><span style="font-family: Helvetica,Arial,sans-serif;">- </span><a
        style="font-family: Helvetica,Arial,sans-serif;"
        href="files/worksOnlyInDiploids.txt">worksOnyInDiploids.txt</a><span
        style="font-family: Helvetica,Arial,sans-serif;">: the list of </span><span
        style="font-weight: bold; font-family:
        Helvetica,Arial,sans-serif;">functions</span><span
        style="font-family: Helvetica,Arial,sans-serif;"> working </span><span
        style="font-weight: bold; font-family:
        Helvetica,Arial,sans-serif;">only for diploid</span><span
        style="font-family: Helvetica,Arial,sans-serif;"> genotypes </span><br
        style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <br style="font-family: Helvetica,Arial,sans-serif;">
      <span style="font-weight: bold; color: rgb(0, 102, 0);
        font-family: Helvetica,Arial,sans-serif;"></span><br
        style="font-family: Helvetica,Arial,sans-serif;">
      <span style="text-decoration: underline; font-family:
        Helvetica,Arial,sans-serif;"></span><br>
      <span style="color: rgb(255, 0, 0);"></span></div>
  </body>
</html>
