<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
  <meta content="text/html; charset=ISO-8859-1"
 http-equiv="content-type">
  <title>news</title>
</head>
<body>
<span style="font-weight: bold;"></span><span style="font-weight: bold;"></span><br>
<br>
<div style="margin-left: 40px;"><big><big><span
 style="font-weight: bold;"></span></big></big><span
 style="font-weight: bold;"></span><span style="font-weight: bold;"></span><span
 style="color: rgb(255, 0, 0);"></span><br>
<br>
<span style="font-weight: bold;"><big><br>
Today</big> </span><span style="font-weight: bold;"></span><span
 style="font-weight: bold;"></span><br>
<span style="font-weight: bold;"></span>Current
stable version of <span style="color: rgb(255, 0, 0);">adegenet is
1.2-4 </span>for R.2.11.1.<br>
You can install the devel version from <a
 href="https://r-forge.r-project.org/scm/?group_id=120">R-Forge</a>.<br>
See the current <a
 href="http://cran.r-project.org/web/packages/adegenet/ChangeLog">ChangeLog</a>
for a list of
all modifications.<br>
Next major release will be 1.3-0, and will include the stable versions
of SeqTrack algorithm, DAPC, and new simulation tools.<br>
<br>
<br>
<span style="font-weight: bold;">19/04/2010 <img
 style="width: 80px; height: 37px;" alt="" src="images/new.png"></span><span
 style="font-weight: bold;"></span><br>
<big>New version of the adegenet tutorial online.</big><span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">
This new version of the tutorial is compatible with the latest release
of adegenet (version 1.2-4). It incorporates a new section on importing
data into adegenet which is more explicit and detailed than the
previous one.</span></span><span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);"><br>
</span></span><span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);"></span></span><br>
<br>
<span style="font-weight: bold;">19/04/2010 <img
 style="width: 80px; height: 37px;" alt="" src="images/new.png"></span><span
 style="font-weight: bold;"></span><br>
<big>New adegenet version (<span style="color: rgb(255, 0, 0);">1.2-4</span>)
has
been released for R-2.10.1 !</big><span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);">
This release corrects a few issues and bugs from previous version, and
includes new material currently under review in scientific journals.
Once 'stable', these new functionalities will give rise to the major
release 1.3-0. Look at </span></span><a
 href="http://cran.r-project.org/web/packages/adegenet/ChangeLog">ChangeLog</a><span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);"> for
more information about this new version.<br>
</span></span><span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);"></span></span><br>
<br>
<span style="font-weight: bold;">15/04/2010 </span><span
 style="font-weight: bold;"><img style="width: 80px; height: 37px;"
 alt="" src="images/new.png"></span><span style="font-weight: bold;"></span><br>
<big>The adegenet forum's archives can now be searched from this
website.</big><span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);"> Look at the 'contact' section for more
information. The next major adegenet release is on its way. It will
include new methods of genetic data analysis, as well as new simulation
systems. Pending work under review, the release of 1.3-0 should be done
within the next two months. Intermediate release correcting minor bugs
may occur meanwhile.<br>
</span></span><span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);"></span></span><br>
<br>
<span style="font-weight: bold;">02/04/2009 </span><span
 style="font-weight: bold;"></span><br>
<big>New adegenet version (<span style="color: rgb(255, 0, 0);">1.2-3</span>)
has
been released for R.2.8.1 !</big><br>
This is a important<span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);"> release, including some significant
improvements
(see </span></span><a href="files/ChangeLog">ChangeLog</a><span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">
file). The most important new feature is that <span
 style="color: rgb(255, 0, 0);">presence/absence </span>data like AFLP
are now handled. A new slot 'type' now distinguishes, in genind and
genpop objects, codominant markers (type="codom") from presence/absence
(type="PA"). Also, <span
 style="color: rgb(255, 0, 0); font-family: monospace;">hybridize</span>
can now perform hybridization between genotypes with any even degree of
ploidy. Another noticeable feature is that SNPs can be obtained from
aligned DNA sequences using <span
 style="color: rgb(255, 0, 0); font-family: monospace;">DNAbin2genind</span>.This
time,
the tutorial has been updated at the same time as the package
release!<br>
<br>
<br>
</span></span><span style="font-weight: bold;">19/12/2008 </span><span
 style="font-weight: bold;"></span><br>
<big>New adegenet version (<span style="color: rgb(255, 0, 0);">1.2-2</span>)
has
been released for R.2.8.0 !</big><br>
<span style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">This
new,
big release implements many new things and corrects a few bugs
(see </span></span><a href="files/ChangeLog">ChangeLog</a><span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">
file). Amongst the biggest news, <span style="font-weight: bold;">several
levels
of ploidy</span> are now handled throughout the package, apart
from a few exceptions. New graphics are available for multivariate
methods: the <span
 style="font-family: monospace; color: rgb(255, 0, 0);">colorplot</span>,
which
can represent genetic diversity using RGB scale, and a <span
 style="font-family: monospace; color: rgb(255, 0, 0);">loadingplot</span>,
which
can be used to assess the most structuring alleles and loci in an
analysis.&nbsp; A new georeferenced genetic dataset is also available
(dataset <span style="color: rgb(255, 0, 0); font-family: monospace;">rupica</span>).
And...
<a href="files/ChangeLog">several other things</a>!<br>
</span></span><br>
<br>
<span style="font-weight: bold;">03/08/2008 </span><span
 style="font-weight: bold;"></span><br>
<big>New adegenet version (<span style="color: rgb(255, 0, 0);">1.2-1</span>)
has
been released for R.2.7.1 !</big><br>
<span style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">This
new
release corrects a few bugs arising in version 1.2-0 (see </span></span><a
 href="files/ChangeLog">ChangeLog</a><span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">
file), especially an issue with the detection of NAs. Documentation was
also improved (have a look at <span
 style="font-family: monospace; color: rgb(255, 0, 0);">?scaleGen</span>).<br>
</span></span><br>
<span style="font-weight: bold;">09/07/2008 </span><span
 style="font-weight: bold;"></span><br>
<big>New adegenet version (<span style="color: rgb(255, 0, 0);">1.2-0</span>)
has
been released for R.2.7.1 !</big><br>
<span style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">This
major
release implements various levels of ploidy and adds some new
features (see </span></span><a href="files/ChangeLog">ChangeLog</a><span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">
file):<br>
&nbsp;&nbsp;&nbsp; - </span><span style="font-family: monospace;">df2genind</span></span>
now handles a 'sep' argument, specifying a separator between the input
alleles. This extends current input capabilities to a wider range of
software.<br>
&nbsp;&nbsp;&nbsp; - <span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);"></span><span
 style="font-family: monospace;">scaleGen</span></span> is a function
similar to 'scale' for genind/genpop objects. Various scalings adapted
to genetic markers are proposed.<br>
&nbsp;&nbsp;&nbsp; - <span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);"></span><span
 style="font-family: monospace;">propTyped</span></span> is a function
giving the the proportion of non-missing data in different ways; this
is particularly useful to detect structuring in missing data.<br>
&nbsp;&nbsp;&nbsp; - various coercion method are now available,
implemented as methods for 'as' and as old S3 coercion (e.g.,
'as.data.frame([myGenindObject])'). Coercion to matrix, data.frame and
ktab (from ade4 package).<br>
&nbsp; <br>
<span style="font-weight: bold;">02/06/2008 </span><span
 style="font-weight: bold;"></span><br>
<big>New adegenet version (<span style="color: rgb(255, 0, 0);">1.1-2</span>)
has
been released for R.2.7.0 !</big><br>
<span style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">This
release
corrects a few bugs and adds some new features (see </span></span><a
 href="files/ChangeLog">ChangeLog</a><span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">
file):<br>
&nbsp;&nbsp;&nbsp; - </span><span style="font-family: monospace;">genind2df</span></span>
now handles the pop argument correctly.<br>
&nbsp;&nbsp;&nbsp; - <span
 style="color: rgb(255, 0, 0); font-family: monospace;">df2genind</span>
does no longer bug when there is an entirely non-typed locus.<br>
&nbsp;&nbsp;&nbsp; - new function <span
 style="color: rgb(255, 0, 0); font-family: monospace;">propShared</span>
is added: computes the proportion of shared alleles among a set of
genotypes (core computations in C).<br>
&nbsp;&nbsp;&nbsp; - significant improvement in the speed of <span
 style="color: rgb(255, 0, 0); font-family: monospace;">genind2df</span>
(more than twice as fast as before).<br>
&nbsp;&nbsp;&nbsp; - improvement of the <span
 style="color: rgb(255, 0, 0);">validity checking</span> for
genind/genpop: <span
 style="color: rgb(255, 0, 0); font-family: monospace;">validObject</span>
now detects duplicates in any kind of names (ind.names, pop.names,
etc.) and prints the corresponding items.<span
 style="font-family: monospace;"><br>
</span><span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);"></span></span><span
 style="font-weight: bold;"><br>
30/04/2008 </span><span style="font-weight: bold;"></span><br>
The new function '<span
 style="font-family: monospace; color: rgb(255, 0, 0);">propShared</span>'
is
available from the devel version of adegenet. It computes the <span
 style="color: rgb(255, 0, 0);">proportion of shared alleles</span>
among a set of genotypes.<br>
<br>
<span style="font-weight: bold;">09/04/2008 </span><span
 style="font-weight: bold;"></span><br>
<big>New adegenet version (<span style="color: rgb(255, 0, 0);">1.1-1</span>)
has
been released for R.2.6.2 !</big><br>
<span style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">This
release
corrects a few bugs and adds some new features:<br>
&nbsp;&nbsp;&nbsp; - <span style="font-family: monospace;"></span></span></span><span
 style="text-decoration: underline;">I/O:</span> <span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);"><span
 style="font-family: monospace;">df2genind</span> no longer fails when
entirely non-type individuals exist.<br>
&nbsp;&nbsp;&nbsp; </span></span><span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);">- <span
 style="text-decoration: underline;">Spatial
genetics:</span></span></span><span style="color: rgb(255, 0, 0);"><span
 style="color: rgb(0, 0, 0);"> <span
 style="font-family: monospace; color: rgb(255, 0, 0);">optimize.monmonier</span>
now computes the 'best' boundary only once instead of twice. The
matched call of the returned object has been fixed. </span></span><span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);"><span
 style="font-family: monospace; color: rgb(255, 0, 0);">monmonier</span>
no longer fails when coordinates are drawn from a regular grid. The
whole code was re-thought</span></span> and optimized for speed.
Monmonier's boundaries can now form loops. Instead of stoping at a
given threshold, it is also possible to ask for a given length of
boundary (argument <span
 style="font-family: monospace; color: rgb(255, 0, 0);">bd.length</span>).
The
function <span style="font-family: monospace; color: rgb(255, 0, 0);">chooseCN</span>
has a new option to return a list of spatial weights defined as the
inverse of spatial distances, at a given exponent.<br>
&nbsp;&nbsp;&nbsp; - <span style="text-decoration: underline;">F
statistics</span>: a wrapper for glob.varcomp has been implemented for
genind objects, through the new function <span
 style="font-family: monospace; color: rgb(255, 0, 0);">fstat</span>.<br>
<span style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">&nbsp;&nbsp;&nbsp;
</span></span>- <span style="text-decoration: underline;">Data
manipulation:</span> The elements of the <span
 style="color: rgb(255, 0, 0); font-family: monospace;">@other</span>
slot are now proceeded
wisely when objects are subsetted using the '[' operator.<br>
<br>
<span style="font-weight: bold;">15/02/2008 </span><span
 style="font-weight: bold;"><br>
</span><span style="font-weight: bold;"></span>A tutorial for the new
version
of adegenet is available from this website.<br>
<br>
<span style="font-weight: bold;">14/02/2008 </span><span
 style="font-weight: bold;"></span><big><span style="font-weight: bold;"><br>
</span><small>New adegenet version (<span style="color: rgb(255, 0, 0);">1.1-0</span>)
has
been released for R.2.6.2.</small></big><br>
<span style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">This
totally
new version includes:<br>
&nbsp;&nbsp;&nbsp; - <span style="text-decoration: underline;">Data
representation:</span> </span></span><span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">S4
classes in replacement of old </span></span><span
 style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);">S3
classes.<br>
&nbsp;&nbsp;&nbsp; - <span style="text-decoration: underline;">Spatial
genetics:</span> the </span></span>spatial
Principal Component Analysis (Jombart et al, submitted to Heredity),
two multivariate spatial tests, and new functionalities for Monmonier's
algorithm.<br>
&nbsp;&nbsp;&nbsp; -<span style="text-decoration: underline;"> I/O:</span>
functions to import data are now 'read' functions; available for
formats of GENETIX, Fstat, Genepop, STRUCTURE and from data.frames of
genotypes. Export from genind to data.frame of genotypes.<br>
&nbsp;&nbsp;&nbsp; - <span style="text-decoration: underline;">Data:</span>
five new simulated geo-referenced datasets<br>
&nbsp;&nbsp;&nbsp; - <span style="text-decoration: underline;">Simulations:</span>
a hybridize function, which creates hybrids from two parent datasets.
Can output to STRUCTURE format.<br>
&nbsp;&nbsp;&nbsp; - <span style="text-decoration: underline;">Data
manipulation:</span> new function to separate data by population.
Accessors to genind and genpop object like with matrices using 'foo[
chosenGenotypes, chosenAlleles]'.<br>
<br>
<span style="font-weight: bold;">07/01/2008 </span><span
 style="font-weight: bold;"></span><span style="font-weight: bold;"><br>
</span>adegenet is now a project on <a
 href="https://r-forge.r-project.org/projects/adegenet/" target="_top">R-Forge</a>.
<span style="font-weight: bold;"><br>
</span><br>
<span style="font-weight: bold;">12/09/2007 </span><span
 style="font-weight: bold;"></span><span style="font-weight: bold;"><br>
</span>Release of a new version of the <span
 style="color: rgb(255, 0, 0);">adegenet tutorial</span> (see documents
section in this site), showing how to find genetic boundaries using
Monmonier's algorithm.<br>
<br>
<span style="font-weight: bold;">12/09/2007 </span><span
 style="font-weight: bold;"></span><span style="font-weight: bold;"><br>
</span>adegenet version <span style="color: rgb(255, 0, 0);">1.0-2</span>
was released for R.2.5.1. <br>
This new version corrects a few bugs (especially in <span
 style="font-family: monospace; color: rgb(255, 0, 0);">optimize.monmonier</span>)
and
adds the following elements:<br>
<br>
- <span style="font-family: monospace;"><span
 style="color: rgb(255, 0, 0);">adegenetWeb</span> </span>is a simple
function opening the adegenet website in the default web browser.<br>
- <span style="font-family: monospace; color: rgb(255, 0, 0);">sim2pop</span><span
 style="color: rgb(255, 0, 0);"> </span>is a dataset obtained by
simulation using the software Easypop. It contains 130 georeferenced
genotypes sampled from two distinct populations.<br>
- <span style="color: rgb(255, 0, 0); font-family: monospace;">monmonier</span>
documentation was improved <span style="color: rgb(255, 0, 0);"></span>by
adding
a genetic example, using <span
 style="font-family: monospace; color: rgb(255, 0, 0);">sim2pop</span><span
 style="color: rgb(255, 0, 0);"> </span>data.<br>
<span style="font-weight: bold;"><br>
30/08/2007 </span><span style="font-weight: bold;"><br>
</span>A new section is added to the <span
 style="color: rgb(255, 0, 0);">download </span>section: <span
 style="color: rgb(255, 0, 0);">patches</span> correcting bugs or
improving functions of the current version of adegenet.<br>
<br>
<span style="font-weight: bold;">25/08/2007 </span><span
 style="font-weight: bold;"></span><span style="font-weight: bold;"><br>
</span>adegenet version 1.0-1 was released for R.2.5.1. <br>
This new version corrects a few bugs (basically in import/export
functions) and adds the following functions:<br>
<br>
- <span style="font-family: monospace;"><span
 style="color: rgb(255, 0, 0);">chooseCN</span> </span>is a simple
interactive tool for choosing and building a connection network from
spatial coordinates. This tool is called by <span
 style="font-family: monospace; color: rgb(255, 0, 0);">monmonier</span><span
 style="color: rgb(255, 0, 0);"> </span>function.<br>
- <span style="font-family: monospace;"><span
 style="color: rgb(255, 0, 0);">monmonier</span>, <span
 style="color: rgb(255, 0, 0);">optimize.</span></span><span
 style="font-family: monospace;"><span style="color: rgb(255, 0, 0);">monmonier</span>,
<span style="color: rgb(255, 0, 0);">plot.</span></span><span
 style="font-family: monospace;"><span style="color: rgb(255, 0, 0);">monmonier</span>
</span>and<span style="font-family: monospace;"> <span
 style="color: rgb(255, 0, 0);">print.</span></span><span
 style="font-family: monospace;"><span style="color: rgb(255, 0, 0);">monmonier</span>
</span>implement the Monmonier algorithm. While not restrained to
genetic data analysis, this method can be used to find genetic
boundaries among individuals or populations based on their allelic
frequencies and spatial coordinates. <br>
<span style="font-weight: bold;"></span><br>
<span style="font-weight: bold;">24/04/2007</span><span
 style="font-weight: bold;"></span><br>
<span style="font-weight: bold;"></span>This website arrives on the web.<br>
<br>
<span style="font-weight: bold;">10/04/2007 <br>
</span>adegenet first release as an
official <span style="text-decoration: underline;"><a
 href="http://www.r-project.org/" target="_top"><img alt=""
 src="images/R.png"
 style="border: 0px solid ; width: 30px; height: 23px;"></a></span>
package for R.2.4.1.<br>
<br>
<br>
</div>
</body>
</html>
